feat(feedback): Add `captureFeedback` method

mydea opened the pull request:
This PR adds a new `captureFeedback` method which is exported from all SDKs.

This method can be used to capture a (new) user feedback from any package.

We follow the same semantics as for other capture methods like `captureException()` or `captureMessage()`: The method returns a string, which is the event id. We do not wait for sending to be successful or not, we just try to send it and return. You can both set an `associatedEventId` (which replaces the event_id of the "old" `captureUserFeedback`), and also pass attachments to send along (which for now are sent as a separate envelope).

For usage in the modal UI, there is still `sendFeedback` which continues to return a promise that resolves with the event ID, or rejects with a meaningful error message if sending fails.  

This also deprecates `captureUserFeedback()`, which is only exported in browser. We cannot remove this yet because `captureFeedback` will only work on newer self-hosted instances, so not all users can easily update. We can/should remove this in v9.

Includes https://github.com/getsentry/sentry-javascript/pull/11626
Fixes 49946

Lms24 commented:
Feel free to merge https://github.com/getsentry/sentry-javascript/pull/11626 into this PR if it makes sense; otherwise we can also merge sequentially :)

Lms24 reviewed (approved):
LGTM!

billyvg reviewed (approved):
LGTM! Thanks for the tests

ryan953 commented on packages/feedback/src/core/sendFeedback.ts:
```diff
@@ -1,100 +1,65 @@
 /**
  * Public API to send a Feedback item to Sentry
  */
 export const sendFeedback: SendFeedback = (
```
Doesn't have to be in this PR, but should we continue to export this publicly? Or just leave it as an internal helper for our own UI?

mydea replied:
that's a good question, I don't have a strong feeling there! I _think_ for backwards compatibility we might as well continue to export it, but I'll defer to whatever you folks think about this 🤔

ryan953 replied:
I think with the new `captureFeedback()` being a thing, this isn't too important for people anymore. If we were going to remove it now would be a fine time. 

It seems like if we keep it, then the docs will be something like:
> There is `captureFeedback()`, and also `sendFeedback()`... the differences are....

OR if we remove it, docs could instead mention:
> There is `captureFeedback()`. You can response by doing something like `client.on('afterSendEvent', (event, response) => ...)` and check that `event.event_id === eventId` 

I'm starting a migration doc to track moving from "feedback beta & v7" into "v8 & GA", so this could fit there too.
https://github.com/getsentry/sentry-javascript/pull/11731

mydea replied:
I'll remove the public export then in a follow up PR 👍

ryan953 commented on packages/core/src/feedback.ts:
```diff
@@ -0,0 +1,61 @@
+
+  const eventId = getCurrentScope().captureEvent(feedbackEvent, hint);
+
+  // For now, we have to send attachments manually in a separate envelope
+  // Because we do not support attachments in the feedback envelope
+  // Once the Sentry API properly supports this, we can get rid of this and send it through the event envelope
+  if (client && transport && dsn && attachments && attachments.length) {
```
@cmanallen @aliu3ntry Is this still true?

cmanallen replied:
You can send whatever you want in an envelope.  You're just capped to three items.  At least, as far as I know.

c298lee replied:
I think Andrew has a PR up but it's not merged yet

cmanallen replied:
I'm confused on how that PR solves the problem?  Are you trying to put attachments into the feedback envelope item?  Or are you trying to send attachments in the same envelope and a feedback item?  These are two different problems.

Let's talk about this on Monday before my PTO.

aliu3ntry replied:
Confirmed with @c298lee -- same envelope as feedback item. Format will look like
<headers>
<feedback item>
<attachment item>
<attachment item>
...

For screenshots, there will be at most 1 attachment item

aliu3ntry replied:
See https://github.com/getsentry/team-replay/issues/393

c298lee replied:
I discussed this with Andrew, but we want to send attachments in the same envelope as the feedback, the attachment will be its own item. We want to use `getCurrentScope().addAttachment()` to add the attachment to the envelope.

mydea replied:
OK, to clarify: This means I will remove the code to add attachments to a separate envelope, and also remove the `attachments` option for the method, instead expecting this to be added to the scope the same way as for any other event? --> This will reduce bundle size and complexity here too, because this should be handled OOTB by `captureEvent` so this should "just work" really. 

This is OK and will work (with up-to-date self-hosted, at least) today?

ryan953 replied:
It looks like https://github.com/getsentry/team-replay/issues/393 hasn't landed yet; but @mydea the SDK should send attachments and feedback together IMO.
Why?
- Sending them in the same payload gives us better guarantees that both parts made it successfully, so we won't get into cases where random attachments are ingested, but no matching feedback exists.
- We'll save a bunch of SDK code
- We can take advantage of being in alpha & beta state for the screenshot feature specifically.... I think it's ok to have a higher min-requirement on the server because turning on screenshots is opt-in right now.

For example, in the docs we could say something like (this is the idea, not the verbatim docs i'd write):
> v8 of the Feedback SDK supports allowing users to include a screenshot with their feedback submission.
> On-Prem users must update to version _${after getsentry/team-replay#393 lands}_ before enabling `showScreenshot: true` in the SDK config.

mydea replied:
Yeah, I agree that this makes sense, we just need to know how long it will take to land this and when it will be on-prem, and decide if we want to hold off on merging this until we know this!

mydea commented:
So I updated this to send attachments in the same envelope! Makes everything easier.
With this change, I also removed the `attachments` option from `captureFeedback()` and `sendFeedback()`, instead leveraging this normally through the event hint. This means you have to do:

```js
Sentry.captureFeedback({ message: 'xx' }, { attachments: [attachment1] });
```

Same as with all other capture methods.

bruno-garcia commented:
The change on ingest to accept feedback and attachments on the same envelope was done through:
* https://github.com/getsentry/relay/pull/3403
 
This change has been live on SaaS and made it to self hosted: 24.4.2
* https://github.com/getsentry/relay/releases/tag/24.4.2

ryan953 commented on packages/feedback/src/core/sendFeedback.ts:
```diff
@@ -1,100 +1,65 @@
-    });
-
-    if (client.emit) {
-      client.emit('beforeSendFeedback', feedbackEvent, { includeReplay: Boolean(includeReplay) });
```
I think we still want to have this beforeSendFeedback hook. This should a) start to capture any buffered replay b) add a breadcrumb to the replay

ryan953 replied:
Long term it'd be better to move into the Actor::onClick event, so we get more replay before they start typing, but that should be a followup because of the quota impact

mydea replied:
We do still emit this, but in `captureFeedback` ;)

aliu3ntry commented:
Assuming this PR sends feedback+attach in same envelope, let me know before you merge so I can turn on the ingest FF! I'ts only set in s4s right now.
- Turned on: accepts same envelope AND separate envs
- Turned off: only accepts separate envs
Or you can make a PR to do so yourself like so: https://github.com/getsentry/sentry-options-automator/pull/1234

> The change on ingest to accept feedback and attachments on the same envelope was done through:
> 
> * [feat(replay): separate feedback from attachments in the same envelope relay#3403](https://github.com/getsentry/relay/pull/3403)
> 
> This change has been live on SaaS and made it to self hosted: 24.4.2
> 
> * https://github.com/getsentry/relay/releases/tag/24.4.2
